account_move_journal_template/models/account_move.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2022 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from odoo import _, fields, models
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    journal_entry_ids = fields.One2many(
+        comodel_name="account.move", inverse_name="journal_entry_id",
+    )
+    journal_entry_id = fields.Many2one(comodel_name="account.move")
+
+    def button_draft(self):
+        res = super().button_draft()
+        if self.journal_entry_ids:
+            self.journal_entry_ids.button_draft()
+        return res
+
+    def button_cancel(self):
+        res = super().button_cancel()
+        if self.journal_entry_ids:
+            self.journal_entry_ids.button_cancel()
+        return res
+
+    def action_view_journal_entries(self):
+        self.ensure_one()
+        return {
+            "name": _("Journal Entries"),
+            "type": "ir.actions.act_window",
+            "res_model": "account.move",
+            "view_mode": "tree,form",
+            "target": "current",
+            "domain": [("id", "in", self.journal_entry_ids.ids)],
+        }
+
+    def post(self):
+        res = super().post()
+        all_journals = self.env["account.move"]
+        for move in self:
+            if move.type not in ["out_invoice", "out_refund"]:
+                continue
+            # Remove previous lines
+            move.journal_entry_ids.line_ids.unlink()
+            product_tmpl_ids = move.mapped(
+                "invoice_line_ids.product_id.product_tmpl_id"
+            )
+            for tmpl in product_tmpl_ids:
```

Comment:
I think it will be much easier if the code follows the structure of the data, something like (pseudocode):
```
DELETE existing extra moves if any (should not happen)
FOR move in self:
   FOR invoice_line IN move.line_ids:
        tmpl = invoice_line.product_id.product_templ_id
        IF NOT tmpl..account_product_move_id:
            continue
         CREATE extra move for product template
         FOR product_move_line IN tmpl..account_product_move_id.line_ids:
             CREATE extra line for product_move_line
         extra_move.post()
```

No need to search for existing extra moves. No need to keep extra moves in a variable. No need to look up the qiuantity for the invoice line product, as you will have this in the invoice_line record...

Reply:
The problem with this one is that we need an extra_move per journal_id of account_product_move_line.